dead/differential_testing.py

```diff
@@ -0,0 +1,202 @@
+"""
+Implementd the core logic of Dead Code Elimination based Differential Testing
+
+e.g., given a diopter.SourceProgram and two diopter.CompilationSettings
+
+if case:= differential_test(program, setting1, setting2):
+    #an interesting case has been found
+"""
+
+from concurrent.futures import ProcessPoolExecutor, as_completed
+from dataclasses import dataclass
+from enum import Enum
+
+from dead_instrumenter.instrumenter import (
+    DCEMarker,
+    InstrumentedProgram,
+    VRMarker,
+    instrument_program,
+)
+from diopter.compiler import CompilationSetting, SourceProgram
+from diopter.generator import CSmithGenerator
+from diopter.sanitizer import Sanitizer
+
+from dead.config import DeadConfig
+from dead.progressbar import progressbar
+
+
+@dataclass(frozen=True, kw_only=True)
+class DifferentialTestingCase:
+    """
+    An instrumented program and two compilation settings, one of which cannot
+    eliminate one or more markers that the other can (and vice versa).
+
+    Attributes:
+        program (InstrumentedProgram):
+            the instrumented program whose markers are missed
+            by one setting and eliminated by the other
+        setting1 (CompilationSetting):
+            the first compilation setting
+        setting2 (CompilationSetting):
+            the second compilation setting
+        markers_only_eliminated_by_setting1 (tuple[DCEMarker | VRMarker, ...]):
+            markers eliminated by setting1 and missed by setting2
+        markers_only_eliminated_by_setting2 (tuple[DCEMarker | VRMarker, ...]):
+            markers eliminated by setting2 and missed by setting1
+    """
+
+    program: InstrumentedProgram
+    setting1: CompilationSetting
+    setting2: CompilationSetting
+    markers_only_eliminated_by_setting1: tuple[DCEMarker | VRMarker, ...]
+    markers_only_eliminated_by_setting2: tuple[DCEMarker | VRMarker, ...]
+
+
+class DifferentialTestingMode(Enum):
```

Comment:
I don't really like the naming scheme, but I couldn't come up with something better. I'm not sure if this mode is needed in the long run.